reconcile/openshift_tekton_resources.py

```diff
@@ -0,0 +1,353 @@
+        if tkn_provider not in tkn_providers:
+            tkn_providers[tkn_provider] = all_tkn_providers[tkn_provider]
+
+        if 'saas_files' not in tkn_providers[tkn_provider]:
+            tkn_providers[tkn_provider]['saas_files'] = []
+
+        tkn_providers[tkn_provider]['saas_files'].append(saas_file)
+
+    return tkn_providers
+
+
+# Create an array of dicts that will be used as args of ri.add_desired
+# This will also add resourceNames inside tkn_providers['namespace']
+# while we are migrating from the current system to this integration
+def fetch_desired_resources(tkn_providers: dict[str, Any]) \
+        -> list[dict[str, Union[str, OR]]]:
+    desired_resources = []
+    for tkn_provider in tkn_providers.values():
+        namespace = tkn_provider['namespace']['name']
+        cluster = tkn_provider['namespace']['cluster']['name']
+        deploy_resources = tkn_provider.get('deployResources',
+                                            DEFAULT_DEPLOY_RESOURCES)
+
+        # a dict with task template names as keys and types as values
+        # we'll use it when building the pipeline object to make sure
+        # that all tasks referenced exist and to be able to set the
+        # the corresponding ['taskRef']['name']
+        task_templates_types = {}
+
+        # desired tasks. We need to keep track of the tasks added in this
+        # namespace, hence we will use this instead of adding data
+        # directly to desired_resources
+        desired_tasks = []
+        for task_template_config in tkn_provider['taskTemplates']:
+            task_templates_types[task_template_config['name']] = \
+                task_template_config['type']
+
+            if task_template_config['type'] == 'onePerNamespace':
+                task = build_one_per_namespace_task(task_template_config)
+                desired_tasks.append(
+                    build_desired_resource(task,
+                                           task_template_config['path'],
+                                           cluster,
+                                           namespace))
+            elif task_template_config['type'] == 'onePerSaasFile':
+                for saas_file in tkn_provider['saas_files']:
+                    task = build_one_per_saas_file_task(
+                        task_template_config, saas_file, deploy_resources)
+                    desired_tasks.append(
+                        build_desired_resource(task,
+                                               task_template_config['path'],
+                                               cluster,
+                                               namespace))
+            else:
+                raise OpenshiftTektonResourcesBadConfigError(
+                    f"Unknown type [{task_template_config['type']}] in tekton "
+                    f"provider [{tkn_provider['name']}]")
+
+        if len(tkn_provider['taskTemplates']) != \
+                len(task_templates_types.keys()):
```

Comment:
Wouldn't it be stricter to `set(tkn_provider['taskTermplates']) == task_templates_types.keys()` (`.keys()` is a set-like thingy)

Reply:
that would not work because `tkn_provider['taskTemplates']` is a complex structure. It's enough with what it's written because if there are duplicates the length will be smaller in the `task_templates_types` dict that has as the task template config as name.